KarhooUISDK/Screens/BookingScreen/QuoteListMVP/QuoteListPresenter.swift

```diff
@@ -0,0 +1,202 @@
+//  Karhoo
+//
+//
+//  Copyright © 2020 Karhoo. All rights reserved.
+//
+
+import KarhooSDK
+import Adyen
+
+final class KarhooQuoteListPresenter: QuoteListPresenter {
+
+    // MARK: - Properties
+
+    private let journeyDetailsManager: JourneyDetailsManager
+    private let quoteService: QuoteService
+    private var fetchedQuotes: Quotes?
+    private var quotesObserver: KarhooSDK.Observer<Quotes>?
+    private var quoteSearchObservable: KarhooSDK.Observable<Quotes>?
+    private var selectedQuoteCategory: QuoteCategory?
+    private var selectedQuoteOrder: QuoteSortOrder = .qta
+    private let quoteSorter: QuoteSorter
+    private let analytics: Analytics
+    private let router: QuoteListRouter
+    var onStateUpdated: ((QuoteListState) -> Void)?
+
+    // MARK: - Lifecycle
+
+    init(
+        router: QuoteListRouter,
+        journeyDetailsManager: JourneyDetailsManager = KarhooJourneyDetailsManager.shared,
+        quoteService: QuoteService = Karhoo.getQuoteService(),
+        quoteSorter: QuoteSorter = KarhooQuoteSorter(),
+        analytics: Analytics = KarhooUISDKConfigurationProvider.configuration.analytics()
+    ) {
+        self.router = router
+        self.journeyDetailsManager = journeyDetailsManager
+        self.quoteService = quoteService
+        self.quoteSorter = quoteSorter
+        self.analytics = analytics
+        journeyDetailsManager.add(observer: self)
+    }
+
+    deinit {
+        journeyDetailsManager.remove(observer: self)
+        quoteSearchObservable?.unsubscribe(observer: quotesObserver)
+    }
+
+    func screenWillAppear() {
+        guard let journeyDetails = journeyDetailsManager.getJourneyDetails() else {
+            assertionFailure("Unable to get data to upload")
+            return
+        }
+        analytics.quoteListOpened(journeyDetails)
+    }
+
+    // MARK: - Endpoints
+
+    func selectedQuoteCategory(_ category: QuoteCategory) {
+        selectedQuoteCategory = category
+        updateViewQuotes()
+    }
+    
+    func didSelectQuoteOrder(_ order: QuoteSortOrder) {
+        selectedQuoteOrder = order
+        updateViewQuotes()
+    }
+
+    func didSelectQuote(_ quote: Quote) {
+        router.routeToQuote(quote)
+    }
+
+    func didSelectQuoteDetails(_ quote: Quote) {
+        // TODO: finish implementation
+    }
+
+    func didSelectCategory(_ category: QuoteCategory) {
+        // TODO: finish implementation
+    }
+
+    // MARK: - Private
+
+    private func quoteSearchSuccessResult(_ quotes: Quotes, journeyDetails: JourneyDetails) {
+        // Checkout component required this data
+        setExpirationDates(of: quotes)
+        if quotes.all.isEmpty && quotes.status != .completed {
+            onStateUpdated?(.loading)
+        } else if quotes.all.isEmpty && quotes.status == .completed {
+            onStateUpdated?(.empty(reason: .noResults))
+        }
+        fetchedQuotes = quotes
+        // Why order is set based on location details?
```

Comment:
We check for a destination first because on iOS the first address to be selected is always the pick-up. If the destination is not nil, then it means that both addresses have been selected. We could add a helper `var areBothAddressesSelected` to replace this `!= nil` comparison for more clarity. At this point the destination should not be nil anyway. So it may be a bit redundant. The logic needs checking, maybe it's just a leftover.

The decision to select a sort order here is based on whether the journey is scheduled or not. If it's an asap booking we let the user sort by price or eta. If it's scheduled for later, then the sort by eta is no longer applicable (the sorter is hidden), so we reset it to be by price.